### ISSUE TITLE
Validate_required() removes/reverts changeset value to original value on error

### ISSUE THREAD

**Issue by ewildgoose** [elixir-ecto/ecto#3121]
The validate_required() function drops the change from the changeset in the case that you try and update it to an error/null value. Basically the original value before applying the change reappears! (see Map.drop(changes, ...)

https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/changeset.ex#L1718

This leads to curious results in many situations:
- If the user is editing an existing record in a web form and tries to blank this required value, then if we ask for the current value of the field using input_value() after calling validate_required(), it's the returned as the original value before editing!
- In contrast if it's a new form, then we of course get the original value which is null
- If the user enters some value which fails some other validation then we get the (incorrect) value that the user entered in the changeset...

This surprised me while writing writing some functions where I need custom presentation of the data in the webform and it would have required a workaround to examine both posted params + ecto changeset to determine what to display

Also I wanted a further function to make a determination to abort and not save the record (say all fields were blank), and this broke in the case of editing an existing record because the original value suddenly re-appeared after this validation (and my changeset was no longer all empty?!)

This appears to be a bug or misfeature. I would recommend this line to be removed since none of the other validations appear to work this way?

**Comment by josevalim** [comment]
It is documented behaviour that `validate_required` removes changes so they do not get validated further. It is recommended further processing code to only continue if the changeset is valid or by checking if `get_change` effectively returns a change.

> This leads to curious results in many situations:

Those situations should not happen in Phoenix forms because Phoenix first reads the value from parameters when re-rendering the form. So please double check. If you have your own forms, they can perform similar action.

> This appears to be a bug or misfeature. I would recommend this line to be removed since none of the other validations appear to work this way?

For better or worse, it is currently documented behaviour and changing it will cause breaking changes. We also don't plan an Ecto v4 ever, so I doubt it will ever change. If you are really unhappy with this behaviour, you can always roll your own validate_required.

Thanks for the report!

**Comment by ewildgoose** [comment]
OK, *NOW* I can see the docs about this... 

My feeling would be that it's a small-ish change, could be voted on for a point change release... Or function renamed, etc. Feels odd that it's the only function that behaves like this? Was there some historic reason to do this?

Regarding Phoenix forms. I'm probably doing things wrong, but I find when you have custom ecto types, eg in my case I have a GPS coord stored in some custom format. I found it complicated to get a human readable display format working using params. Can't remember the specifics, but sometimes I had native and sometimes text. Most reliable is to read the value from the changeset and format it for display (which goes wrong if I used validate_required)

Also I wanted the functionality that it was an error if either lat/lon were missing, but if both were missing it was valid (and the record deleted). I couldn't implement behaviour this using validate_required, so I wrote my own validate function instead.

Anyway, I would vote to change it somehow, but thanks for pointing out I'm an idiot for missing this in the docs - had read that section multiple times trying to understand... sigh.. Bold?
